// chardet/lib/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const utf8_1 = __importDefault(require("./encoding/utf8"));
const unicode = __importStar(require("./encoding/unicode"));
const mbcs = __importStar(require("./encoding/mbcs"));
const sbcs = __importStar(require("./encoding/sbcs"));
const iso2022 = __importStar(require("./encoding/iso2022"));
let fs;
const loadFs = () => {
    if (typeof module === 'object' && typeof module.exports === 'object') {
        fs = fs ? fs : require('fs');
        return fs;
    }
    throw new Error('File system is not available');
};
var recognisers = [
    new utf8_1.default(),
    new unicode.UTF_16BE(),
    new unicode.UTF_16LE(),
    new unicode.UTF_32BE(),
    new unicode.UTF_32LE(),
    new mbcs.sjis(),
    new mbcs.big5(),
    new mbcs.euc_jp(),
    new mbcs.euc_kr(),
    new mbcs.gb_18030(),
    new iso2022.ISO_2022_JP(),
    new iso2022.ISO_2022_KR(),
    new iso2022.ISO_2022_CN(),
    new sbcs.ISO_8859_1(),
    new sbcs.ISO_8859_2(),
    new sbcs.ISO_8859_5(),
    new sbcs.ISO_8859_6(),
    new sbcs.ISO_8859_7(),
    new sbcs.ISO_8859_8(),
    new sbcs.ISO_8859_9(),
    new sbcs.windows_1251(),
    new sbcs.windows_1256(),
    new sbcs.KOI8_R()
];
exports.detect = (buffer) => {
    const matches = exports.analyse(buffer);
    return matches.length > 0 ? matches[0].name : null;
};
exports.analyse = (buffer) => {
    var fByteStats = [];
    for (var i = 0; i < 256; i++)
        fByteStats[i] = 0;
    for (var i = buffer.length - 1; i >= 0; i--)
        fByteStats[buffer[i] & 0x00ff]++;
    var fC1Bytes = false;
    for (var i = 0x80; i <= 0x9f; i += 1) {
        if (fByteStats[i] != 0) {
            fC1Bytes = true;
            break;
        }
    }
    var context = {
        fByteStats: fByteStats,
        fC1Bytes: fC1Bytes,
        fRawInput: buffer,
        fRawLength: buffer.length,
        fInputBytes: buffer,
        fInputLen: buffer.length
    };
    var matches = recognisers
        .map((rec) => {
        return rec.match(context);
    })
        .filter((match) => {
        return !!match;
    })
        .sort((a, b) => {
        return b.confidence - a.confidence;
    });
    return matches;
};
exports.detectFile = (filepath, opts = {}) => new Promise((resolve, reject) => {
    var fd;
    const fs = loadFs();
    var handler = (err, buffer) => {
        if (fd) {
            fs.closeSync(fd);
        }
        if (err) {
            reject(err);
        }
        else {
            resolve(exports.detect(buffer));
        }
    };
    if (opts && opts.sampleSize) {
        fd = fs.openSync(filepath, "r");
        const sample = Buffer.allocUnsafe(opts.sampleSize);
        fs.read(fd, sample, 0, opts.sampleSize, null, (err) => {
            handler(err, sample);
        });
        return;
    }
    fs.readFile(filepath, handler);
});
exports.detectFileSync = (filepath, opts = {}) => {
    const fs = loadFs();
    if (opts && opts.sampleSize) {
        var fd = fs.openSync(filepath, "r"), sample = Buffer.allocUnsafe(opts.sampleSize);
        fs.readSync(fd, sample, 0, opts.sampleSize);
        fs.closeSync(fd);
        return exports.detect(sample);
    }
    return exports.detect(fs.readFileSync(filepath));
};
